File: 13_stacks_and_queues/balanced_paranthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        close_map = {")": "(", "}": "{", "]": "["}

        for i in s:
            if i not in close_map:
                stack.append(i)
                continue

            if not stack:
                return False
            if stack.pop() != close_map[i]:
                return False

        return stack == []

File: 13_stacks_and_queues/balanced_paranthesis.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let close_map: HashMap<char, char> = HashMap::from([(']', '['), (')', '('), ('}', '{')]);
        let mut stack: Vec<char> = vec![];

        for c in s.chars() {
            if !close_map.contains_key(&c) {
                stack.push(c);
                continue;
            }

            if stack.len() == 0 || stack.pop().unwrap() != *close_map.get(&c).unwrap() {
                return false;
            }
        }

        stack.len() == 0
    }
}
